lib/ansible/plugins/action/reboot.py

```diff
@@ -0,0 +1,277 @@
+            result['rebooted'] = False
+            result['msg'] = "Shutdown command failed. Error was %s, %s" % (
+                to_native(reboot_result['stdout'].strip()), to_native(reboot_result['stderr'].strip()))
+            return result
+
+        result['failed'] = False
+
+        # Get the original connection_timeout option var so it can be reset after
+        result['connection_timeout_orig'] = None
+        try:
+            result['connection_timeout_orig'] = self._connection.get_option('connection_timeout')
+        except AnsibleError:
+            display.debug("%s: connect_timeout connection option has not been set" % self._task.action)
+
+        post_reboot_delay = int(self._task.args.get('post_reboot_delay', self._task.args.get('post_reboot_delay_sec', self.DEFAULT_POST_REBOOT_DELAY)))
+        if post_reboot_delay < 0:
+            post_reboot_delay = 0
+
+        if post_reboot_delay != 0:
+            display.vvv("%s: waiting an additional %d seconds" % (self._task.action, post_reboot_delay))
+            time.sleep(post_reboot_delay)
+
+        return result
+
+    def validate_reboot(self, before_uptime, connection_timeout_orig):
+        display.debug('%s: Validating reboot' % self._task.action)
+        result = {}
+
+        try:
+            # keep on checking system uptime with short connection responses
+            reboot_timeout = int(self._task.args.get('reboot_timeout', self._task.args.get('reboot_timeout_sec', self.DEFAULT_REBOOT_TIMEOUT)))
+            connect_timeout = self._task.args.get('connect_timeout', self._task.args.get('connect_timeout_sec', self.DEFAULT_CONNECT_TIMEOUT))
+            self.do_until_success_or_timeout(self.check_uptime, reboot_timeout, before_uptime, action_desc="uptime check")
+
+            if connect_timeout:
+                # reset the connection to clear the custom connection timeout
+                try:
+                    self._connection.set_option("connection_timeout", connect_timeout)
+                    self._connection._reset()
+                except (AnsibleError, AttributeError) as e:
+                    display.debug("Failed to reset connection_timeout back to default: %s" % to_text(e))
+
+            # finally run test command to ensure everything is working
+            # FUTURE: add a stability check (system must remain up for N seconds) to deal with self-multi-reboot updates
+            self.do_until_success_or_timeout(self.run_test_command, reboot_timeout, before_uptime, action_desc="post-reboot test command")
+
+            result['rebooted'] = True
+            result['changed'] = True
+
+        except TimedOutException as toex:
+            result['failed'] = True
+            result['rebooted'] = True
+            result['msg'] = to_text(toex)
+            return result
+
+        return result
+
+    def run(self, tmp=None, task_vars=None):
+        self._supports_check_mode = True
+        self._supports_async = True
+
```

Comment:
Probably should also check and fail for `self._connection.transport == 'local'` - no need to go nutty trying to figure out if they're rebooting the controller via a more circuitous route, but at least cover the easy one.